Improve dotnet nuget push to Azure DevOps Artifacts documentation (#29917)
dotnet/docs: docs/core/tools/dotnet-nuget-push.md

- **`-k|--api-key <API_KEY>`**

  The API key for the server.

>   - Azure DevOps Artifacts feeds doesn't accept PAT(personal access tokens) passed directly in cli as `API_KEY`. For dev box, you need have the [cred provider](https://github.com/microsoft/artifacts-credprovider) installed and that would work for both push and download (see examples section). For CI, you need use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) task with [NuGet service connection](https://docs.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#nuget-service-connection) for pushing to Azure DevOps Artifacts.

Reviewer: I think this is slightly out of place here. 

IMO the description should match what the tooling provides. 

Pushing to Azure DevOps can be its own section.

>   - Azure DevOps Artifacts feeds doesn't accept PAT(personal access tokens) passed directly in cli as `API_KEY`. For dev box, you need have the [cred provider](https://github.com/microsoft/artifacts-credprovider) installed and that would work for both push and download (see examples section). For CI, you need use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) task with [NuGet service connection](https://docs.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#nuget-service-connection) for pushing to Azure DevOps Artifacts.

Reviewer 2: The abbreviation should be in parentheses and then you can use the abbreviation later now that its been spelled out once.

```suggestion
  - Azure Artifacts feeds do not accept a personal access token (PAT) specified by this command-line argument. When using a local development environment, you must have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section). For hosted builds that interfact with Azure Artifacts feeds, you must use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task.
```

  ```dotnetcli
  dotnet nuget push "*.nupkg" -s c:\mydir
  ```
  

> - In the example below pushing "Foo" version "5.0.2" to Azure DevOps Artifacts from dev box, here AZ is just a placeholder for ApiKey, this prevents authentication fail prematurely, in order to authentication to work you need to install [cred provider](https://github.com/microsoft/artifacts-credprovider). Below command trigger open Cred Provider window if authentication is necessary, it's suitable for pushing from dev box, but not for CI.

Reviewer: Similarly to above, I think we should have a dedicated section for azure devops.

Reviewer: Feels like we should be referencing an azure devops page with instructions for this instead.

Does Azure Devops have documentation on how to handle pushing at all?

Reviewer 3: AzDO does have this page on publishing packages: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml

I don't feel 100% comfortable putting feed specific instructions in our push docs that's generic for all feeds. We certainly do have a lot of Azure Artifacts customers, and it is a Microsoft service. And if customers come to this help page, but don't find Azure Artifacts' docs page, then instructions here can help them be successful.

On the other hand, GitHub and nuget.org and are also owned by Microsoft. Should we put instructions for those NuGet servers in this doc too?  How about non-Microsoft servers like MyGet, GitLabs, TeamCity, Artifactory, and who knows how many more?  I might be just too idealistic, but I don't feel like server specific information belong in the generic tool docs.

Reviewer 3: Also, isn't the .NET SDK and CLI supposed to be intentionally platform agnostic?  To help dispel the myth that .NET is for Microsoft platforms only. I'm not sure that documenting special instructions for Azure Artifacts only helps maintain that independence.

Author: > AzDO does have this page on publishing packages: https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml
> 
This is opposite direction (Azdo -> xxxx) of what I added here. We can't use Azdo task to push nuget to Azdo from local or Github actions.

> On the other hand, GitHub and nuget.org and are also owned by Microsoft. Should we put instructions for those NuGet servers in this doc too? How about non-Microsoft servers like MyGet, GitLabs, TeamCity, Artifactory, and who knows how many more? I might be just too idealistic, but I don't feel like server specific information belong in the generic tool docs.

Yes, you concern is valid. Still most like customer come to this doc page due to search engines so having this info relieve customer frustration. So far, MyGet, GPR (I manually tested them) and most nuget server implemenet uses `apikey`, but Azdo ignores actual value of `apikey` passed, still requires passing something (in this case `AZ`). Considering quite many customer upvoted on https://github.com/NuGet/Home/issues/7792 I believe we need to document it, I'm not aware of any other package server has a quirk like Azdo.

Reviewer: > I believe we need to document it, I'm not aware of any other package server has an apikey quirk like Azdo.

I agree that we should document it, but it doesn't have to be a section in the push description page.

Reviewer 3: > This is opposite direction (Azdo -> xxxx) of what I added here.

I copied the wrong URL, but if you looked at the table of contents on the left, it was the article immediately underneath the one I linked: [Publish a NuGet package using the command line](https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops). To me, that is exactly what you're documenting. How to publish a nuget package using the CLI, to an Azure Artifacts feed.

Author: Created link to example in Azure Artifacts examples.

> - In the example below pushing "Foo" version "5.0.2" to Azure DevOps Artifacts from dev box, here AZ is just a placeholder for ApiKey, this prevents authentication fail prematurely, in order to authentication to work you need to install [cred provider](https://github.com/microsoft/artifacts-credprovider). Below command trigger open Cred Provider window if authentication is necessary, it's suitable for pushing from dev box, but not for CI.

Reviewer 2: ```suggestion
- In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.
```

  ```dotnetcli

>   dotnet nuget push Foo.5.0.2.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ

Reviewer 2: ```suggestion
  dotnet nuget push MyPackage.5.0.2.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```

  ```
  

> - In the example below "Foo" version "1.1.5" nupkg to Azure DevOps Artifacts from CI, here AZ is just a placeholder for ApiKey, this prevents authentication fail prematurely. You need to setup [NuGet Authenticate task](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops) with [NuGet service connection](https://docs.microsoft.com/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#nuget-service-connection) for authenticate with external Azure DevOps Artifacts server.

Reviewer 2: ```suggestion
- In the example below, pushing "MyPackage" version "1.1.5" to an Azure Artifacts feed from a hosted build, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires your pipeline to use the [NuGet Authenticate task](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops).
```

  ```bashcli
    - task: NuGetAuthenticate@1
      inputs:
        nuGetServiceConnections: MyServiceConnection_ExternalServer
        
    - bash: |

>         dotnet build foo/foo.csproj -c Release

Reviewer 2: ```suggestion
        dotnet build mypackage/mypackage.csproj -c Release
```

>         dotnet pack foo/foo.csproj /property:PackageVersion=1.1.5 -o nupkgs -c Release

Reviewer 2: ```suggestion
        dotnet pack mypackage/mypackage.csproj /property:PackageVersion=1.1.5 -o nupkgs -c Release
```

>         dotnet nuget push nupkgs/foo.1.1.5.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ

Reviewer 2: ```suggestion
        dotnet nuget push nupkgs/mypackage.1.1.5.nupkg --source https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json -k AZ
```
